### DISCUSSION TITLE
Trapping SSE

### DISCUSSION THREAD

**Discussion by Mashiane** [pocketbase/pocketbase#1173] [Q&A]
Hi

I need some help please.

1. I am able to execute a GET for /api/realtime and are able to get a client id.

I am trying to no avail to make the next step work, i.e. executing a POST to /api/realtime for subcriptions.

I am either getting a missing client id or a bad request and thus confused.

Can you please be kind and provide a working example as part of your docs for the whole cycle please?

Also, how does one trap the SSE events for create, update, delete when the subscriptions are done via the API and not directly javascript?

Thanks in advance.

**Comment by ganigeorgiev** [comment]
> I am trying to no avail to make the next step work, i.e. executing a POST to /api/realtime for subcriptions.

To submit a realtime subscription you'll need 2 things:
- the `clientId` from the `PB_CONNECT` event
- your subscription topic(s) (eg. `["COLLECTION_NAME1", "COLLECTION_NAME2/RECORD_ID"]`)

(_you can find more API details in https://pocketbase.io/docs/api-realtime/_)


---

> Can you please be kind and provide a working example as part of your docs for the whole cycle please?

You can find some general SSE flow instructions in https://github.com/pocketbase/pocketbase/discussions/308#discussioncomment-3389399.

The realtime documentation was also recently updated - https://pocketbase.io/docs/api-realtime/.

In short, the realtime events require 2 requests:

1. `GET /api/realtime` - this is a long lived HTTP connection **where you'll receive all SSE event messages**.

2. `POST /api/realtime` - this is a regular/short lived HTTP request that just registers the subscriptions/topics the client wants to be notified to. The authorization also happens during the subscriptions submission (aka. you can send `Authorization: TOKEN` header with this request and the associated `clientId` will be linked to the "auth session").

---

> Also, how does one trap the SSE events for create, update, delete when the subscriptions are done via the API and not directly javascript?

I'm not sure that I understand the question. We use the standard SSE message format and each message has at least 2 fields:

```
event:collection1                          // the subscription/topic name
data:{"action": "create", "record": {...}} // serialized json with the event data
```

**Comment by Mashiane** [comment]
Ok thanks for the heads up, I think I understand the flow, sadly on my side i'm always getting a 404 or 400 on the Post, even if I do everything as explained and use the clientid i got from the SSE. I am not adding any authorization as yet.

ResponseError. Reason: Not Found, Response: {"code":404,"message":"Missing or invalid client id.","data":{}}

ResponseError. Reason: Bad Request, Response: {"code":400,"message":"Something went wrong while processing your request.","data":{"clientId":{"code":"validation_required","message":"Cannot be blank."}}}

**Reply by ganigeorgiev** [reply]
The last error message indicates that you are not sending the `clientId` with the request.

Make sure that the request has the correct content type header (eg. `Content-Type: application/json`).

**Reply by ganigeorgiev** [reply]
You can also start the executable with `--debug` and check the realtime logs in the console.

**Reply by Mashiane** [reply]
Thank you with the debug flag and your help, I can now see what is happening. I'm using a java library that does http requests. That is where the problem is as it closes the request prematurely.
